Record usage event when traces logged to MLflow server

dbczumar opened the pull request:
<details><summary>&#x1F6E0 DevTools &#x1F6E0</summary>
<p>

[![Open in GitHub Codespaces](https://github.com/codespaces/badge.svg)](https://codespaces.new/dbczumar/mlflow/pull/18822?quickstart=1)

#### Install mlflow from this PR

```
# mlflow
pip install git+https://github.com/mlflow/mlflow.git@refs/pull/18822/merge
# mlflow-skinny
pip install git+https://github.com/mlflow/mlflow.git@refs/pull/18822/merge#subdirectory=libs/skinny
```

For Databricks, use the following command:

```
%sh curl -LsSf https://raw.githubusercontent.com/mlflow/mlflow/HEAD/dev/install-skinny.sh | sh -s pull/18822/merge
```

</p>
</details>

<!-- Remove unused checkboxes except for the patch release section at the bottom -->

### Related Issues/PRs

<!-- Uncomment 'Resolve' if this PR can close the linked items. -->
<!-- Resolve --> #xxx

### What changes are proposed in this pull request?

Record usage event when traces logged to MLflow server

### How is this PR tested?

- [ ] Existing unit/integration tests
- [X] New unit/integration tests
- [ ] Manual tests

<!-- Attach code, screenshot, video used for manual testing here. -->

### Does this PR require documentation update?

- [ ] No. You can skip the rest of this section.
- [X] Yes. I've updated:
  - [ ] Examples
  - [ ] API references
  - [ ] Instructions

### Release Notes

#### Is this a user-facing change?

- [X] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

<!-- Details in 1-2 sentences. You can just refer to another PR with a description if this PR is part of a larger change. -->

#### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/evaluation`: MLflow model evaluation features, evaluation metrics, and evaluation workflows
- [ ] `area/gateway`: MLflow AI Gateway client APIs, server, and third-party integrations
- [ ] `area/prompts`: MLflow prompt engineering features, prompt templates, and prompt management
- [ ] `area/tracing`: MLflow Tracing features, tracing APIs, and LLM tracing functionality
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/build`: Build and test infrastructure for MLflow
- [X] `area/docs`: MLflow documentation pages

<!--
Insert an empty named anchor here to allow jumping to this section with a fragment URL
(e.g. https://github.com/mlflow/mlflow/pull/123#user-content-release-note-category).
Note that GitHub prefixes anchor names in markdown with "user-content-".
-->

<a name="release-note-category"></a>

#### How should the PR be classified in the release notes? Choose one:

- [X] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

#### Should this PR be included in the next patch release?

`Yes` should be selected for bug fixes, documentation updates, and other small changes. `No` should be selected for new features and larger changes. If you're unsure about the release classification of this PR, leave this unchecked to let the maintainers decide.

<details>
<summary>What is a minor/patch release?</summary>

- Minor release: a release that increments the second part of the version number (e.g., 1.2.0 -> 1.3.0).
  Bug fixes, doc updates and new features usually go into minor releases.
- Patch release: a release that increments the third part of the version number (e.g., 1.2.0 -> 1.2.1).
  Bug fixes and doc updates usually go into patch releases.

</details>

<!-- Do not modify or remove any text inside the parentheses. Keep both checkboxes below. -->

- [X] Yes (this PR will be cherry-picked and included in the next patch release)
- [ ] No (this PR will be included in the next minor release)

dbczumar committed:
Add telemetry event for OTEL trace ingestion

Emit OtelTraceReceivedEvent whenever a root span is ingested via the
OTEL API endpoint. The event captures the trace_id and experiment_id
for tracking trace ingestion metrics.

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
Add span_count to OtelTraceReceivedEvent

Track the number of spans in each trace for telemetry metrics.

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
Remove PII from OtelTraceReceivedEvent

Remove trace_id and experiment_id as they could contain PII. Only track
span_count as an aggregate metric.

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
Add test for OtelTraceReceivedEvent telemetry

Test verifies that telemetry event is emitted when traces are ingested
via OTEL API with correct span_count parameter.

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
Add MLflow client detection to OTEL trace telemetry

This commit enhances the OtelTraceReceivedEvent telemetry to identify
traces originating from the MLflow client by checking for:
- User-Agent: mlflow-python-client/*
- X-MLflow-Client-Version header

Changes:
- Modified get_otlp_exporter to include MLflow request headers when
  creating the OTLP span exporter
- Updated export_traces endpoint to capture User-Agent and
  X-MLflow-Client-Version headers
- Added from_mlflow_client field to OtelTraceReceivedEvent
- Added comprehensive test coverage for MLflow client header detection

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
Replace hardcoded header test with true E2E test

Replace the test that hardcoded headers in the request with a true
end-to-end test that uses @mlflow.trace to verify MLflow's OTLP
exporter automatically sends client identification headers.

The new test uses @mlflow.trace decorator and verifies from_mlflow_client
is True without hardcoding any headers.

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
Fix test to use standard MLflow client, not OTLP export

The test was incorrectly configuring OTLP export. Changed to use
standard MLflow client configuration with @mlflow.trace to verify
that client headers are sent automatically without any special config.

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
Consolidate telemetry tests into single test with two cases

Combined test_mlflow_client_sends_headers_to_otel_endpoint and
test_otel_trace_received_telemetry_event into one test with:
- Test case 1: Standard MLflow client -> from_mlflow_client=True
- Test case 2: Direct protobuf without headers -> from_mlflow_client=False

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
Remove verbose assertion message from telemetry test

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
Split telemetry test into separate test functions

Split the single test into two separate test functions:
- test_otel_trace_received_telemetry_from_mlflow_client: Tests @mlflow.trace with standard MLflow client
- test_otel_trace_received_telemetry_from_external_client: Tests direct protobuf request without headers

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
Replace from_mlflow_client boolean with source enum

Replace the from_mlflow_client boolean field with a source string field using an OtelTraceSource enum.
The enum has two values:
- MLFLOW: Traces from the MLflow client (identified by User-Agent or X-MLflow-Client-Version headers)
- UNKNOWN: Traces from external OpenTelemetry clients

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
Move OtelTraceSource enum closer to OtelTraceReceivedEvent

Move the OtelTraceSource enum definition to be right before OtelTraceReceivedEvent
for better code organization and locality.

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
Remove span_count field from OtelTraceReceivedEvent telemetry

- Remove span_count from OtelTraceReceivedEvent.parse() method
- Remove span_count from telemetry emission in otel_api.py
- Remove span_count assertions from both telemetry tests
- Move telemetry imports to top of test file per code style

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
Use constants for header names and user agent prefix in OTEL API

- Add _USER_AGENT_HEADER, _CLIENT_VERSION_HEADER, and _MLFLOW_CLIENT_USER_AGENT_PREFIX constants
- Replace hardcoded header strings with constants in export_traces function
- Replace hardcoded user agent prefix check with constant
- Improves maintainability and consistency with existing code patterns

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
Remove useless inline comments from telemetry tests

- Remove obvious comments that don't add value
- Code is self-explanatory without the comments
- Keep docstrings which provide high-level test context

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
Simplify root span tracking to use set instead of dict

- Change root_spans_by_trace_id dict to root_trace_ids set
- Only need to track trace IDs, not store actual span objects
- Reduces memory usage since span_count is no longer emitted

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
Import header constants instead of duplicating them

- Import _USER_AGENT and _CLIENT_VERSION from default_request_header_provider
- Remove duplicated header name constants
- Keep _MLFLOW_CLIENT_USER_AGENT_PREFIX since it's not defined elsewhere
- Follows DRY principle and maintains single source of truth

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
Rename MLFLOW to MLFLOW_CLIENT in OtelTraceSource enum

- More specific and clearer naming
- MLFLOW_CLIENT better describes traces from MLflow Python client
- Updated enum value, API code, and tests

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
Simplify test by using protobuf constructor arguments

- Use nested constructor calls instead of creating intermediate variables
- Reduced from ~40 lines to ~35 lines
- More concise while remaining readable

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
Rename OtelTraceSource to TraceSource and OtelTraceReceivedEvent to TraceReceivedByServerEvent

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
Define user agent prefix in request header provider and import from there

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
Use only X-MLflow-Client-Version header to identify MLflow client, remove user agent check

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
Remove useless comments

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
fix

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
Rename root_trace_ids to trace_ids and add comment explaining root span assumption

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
fix

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
fix

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
Use User-Agent header to identify MLflow Python client and rename to MLFLOW_PYTHON_CLIENT

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
Remove MLflow headers from OTLP exporter

The OTLP exporter is used by MLflow clients to export traces to external
OTLP collectors. These collectors don't need (and may reject) MLflow-specific
headers like User-Agent and X-MLflow-Client-Version.

The telemetry tracking feature works on the SERVER side (otel_api.py endpoint)
where the MLflow client sends headers with HTTP requests. The client-side
OTLP exporter should not include these headers when exporting to external
collectors.

This fixes test failures in test_export_to_otel_collector where spans were
not being received by the external OTLP collector.

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
fix

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
Fix MDX parsing error in usage-tracking.mdx

Add missing backticks around JSON example for trace_received_by_server
event to prevent MDX parser from treating it as JSX expression.

Signed-off-by: dbczumar <corey.zumar@databricks.com>

serena-ruan commented on mlflow/server/otel_api.py:
```diff
@@ -118,9 +126,16 @@ async def export_traces(
+                        # Only add when we find a root span
+                        # (OTEL traces typically have one root span)
+                        trace_ids.add(trace_id)
+                        break
```
This is just `trace_ids = spans_by_trace_id.keys()`?

serena-ruan replied:
I see, that makes sense!

serena-ruan commented on mlflow/server/otel_api.py:
```diff
@@ -141,4 +151,18 @@ async def export_traces(
+                {
+                    "source": trace_source,
+                },
+            )
```
Emm as long as the server is running it should be the same session ID for all traces?

dbczumar replied:
Sorry, you're right - I misinterpreted. I think it's one session ID per handler process, but that's fine :)

dbczumar committed:
Rename trace_ids to completed_trace_ids for clarity

The variable only contains trace IDs where a root span was logged,
indicating completed traces. This better reflects its purpose.

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
Clarify comment about root span indicating completed trace

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
Pluralize TraceReceivedByServerEvent to TracesReceivedByServerEvent

- Rename event class from TraceReceivedByServerEvent to TracesReceivedByServerEvent
- Update event name from trace_received_by_server to traces_received_by_server
- Add count field to event schema to track number of completed traces
- Emit single event with count instead of one event per trace
- Update all tests to use new event name and verify count field

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
Add validation to require count field in TracesReceivedByServerEvent

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
Update test to verify count field with 2 completed traces

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
Update usage tracking docs for TracesReceivedByServerEvent

- Rename event from trace_received_by_server to traces_received_by_server
- Update description to reflect plural traces and count parameter
- Update example to show count field

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
Remove unnecessary intermediate variable in TracesReceivedByServerEvent.parse

Signed-off-by: dbczumar <corey.zumar@databricks.com>

serena-ruan commented on mlflow/telemetry/events.py:
```diff
@@ -320,3 +321,24 @@ def parse(cls, arguments: dict[str, Any]) -> dict[str, Any] | None:
+        return {
+            "source": source.value if isinstance(source, TraceSource) else source,
+            "count": arguments["count"],
+        }
```
```suggestion
```
We don't need this since TracesReceivedByServerEvent is not used in decorator, parse method is only used when the event is used as `@record_usage_event(event_name)` :)

serena-ruan reviewed (approved):
LGTM once https://github.com/mlflow/mlflow/pull/18822/files#r2532433187 is addressed! Thanks for adding this, will be super helpful! :D

dbczumar committed:
Update mlflow/telemetry/events.py

Co-authored-by: Serena Ruan <82044803+serena-ruan@users.noreply.github.com>
Signed-off-by: Corey Zumar <39497902+dbczumar@users.noreply.github.com>

dbczumar commented:
/autoformat

dbczumar committed:
fix

Signed-off-by: dbczumar <corey.zumar@databricks.com>
